pkg/util/resource_parent_map.go

```diff
@@ -0,0 +1,220 @@
+	r.RLock()
+	defer r.RUnlock()
+
+	parents, ok := rpmInstance.childToParentsMap[child]
+	if !ok {
+		return nil
+	}
+
+	// Find parents of particular kind
+	matchingParents := []KubeResource{}
+	for i := range parents {
+		if parents[i].Kind == parentKind {
+			matchingParents = append(matchingParents, parents[i])
+		}
+	}
+
+	// Will return empty slice if nothing found
+	return matchingParents
+}
+
+// GetChildrenOfKind ...
+// WARNING: this function has not been tested. It is likely to not work as desired.
+func (r *ResourceParentsMap) GetChildrenOfKind(parent KubeResource, childKind string) []KubeResource {
+	r.RLock()
+	defer r.RUnlock()
+
+	// First get all of the child keys of the appropriate kind
+	matchKeys := []KubeResource{}
+	for childKey := range rpmInstance.childToParentsMap {
+		if childKey.Kind == childKind {
+			matchKeys = append(matchKeys, childKey)
+		}
+	}
+
+	// Next filter down to only children with the requested parent
+	childrenWithCorrectParent := make(map[KubeResource]int)
+	for i := range matchKeys {
+		parents, ok := rpmInstance.childToParentsMap[matchKeys[i]]
+		if !ok {
+			panic("GetChildrenOfKind - expected child key not found")
+		}
+		// Check if desired parent is attached to child
+		for j := range parents {
+			if parents[j] == parent {
+				// Using a map here as a quasi-set
+				childrenWithCorrectParent[matchKeys[i]] = 1
+			}
+		}
+	}
+
+	// Build the final list of correct parents to return
+	childrenOfKind := []KubeResource{}
+	for child := range childrenWithCorrectParent {
+		childrenOfKind = append(childrenOfKind, child)
+	}
+
+	return childrenOfKind
+}
+
+// AddChildToParent ...
+func (r *ResourceParentsMap) AddChildToParent(child KubeResource, parent KubeResource) {
+	r.Lock()
+	defer r.Unlock()
+
+	parents, ok := rpmInstance.childToParentsMap[child]
+
+	// Create parent list on child if it doesn't exist
+	if !ok {
+		rpmInstance.childToParentsMap[child] = []KubeResource{}
+		parents = rpmInstance.childToParentsMap[child]
+	}
+
+	// Don't add a duplicate parent refs to a child
+	for i := range parents {
+		if parents[i] == parent {
+			return
+		}
+	}
+
+	// Add the new parent to the child resource
+	parents = append(parents, parent)
+	rpmInstance.childToParentsMap[child] = parents
+}
+
+// DeleteChildFromParent ...
+func (r *ResourceParentsMap) DeleteChildFromParent(child KubeResource, parent KubeResource) {
+	r.Lock()
+	defer r.Unlock()
+
+	childParents, ok := rpmInstance.childToParentsMap[child]
+	if !ok {
+		return
+	}
+
+	for i := range childParents {
+		if childParents[i] == parent {
+			childParents = append(childParents[:i], childParents[i+1:]...)
+			rpmInstance.childToParentsMap[child] = childParents
+			return
+		}
+	}
+}
+
+// MapChildToParents can be called from handlerFuncs to lookup all parents of a child
+func MapChildToParents(a handler.MapObject, childKind string, parentKind string) []reconcile.Request {
```

Comment:
Seems like this could (_should_) go in a base type for all the reference `handlers`.